subprojects/functional/src/main/java/org/gradle/internal/Deferrable.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * An invocation to be executed at most once, but one that can be deferred.
+ *
+ * @param <T> The type which will be computed.
+ */
+public interface Deferrable<T> {
+
+    /**
+     * The result of the invocation when it is already available.
+     */
+    Optional<T> getCompleted();
+
+    /**
+     * Obtain the result of the invocation, either by returning the already computed result or by computing it synchronously.
+     *
+     * A result is only calculated once.
+     */
+    T completeAndGet();
+
+    /**
+     * Maps the result of the invocation via a mapper.
+     *
+     * @param mapper An inexpensive function on the result.
+     */
+    default <U> Deferrable<U> map(Function<? super T, U> mapper) {
+        return new Deferrable<U>() {
+            @Override
+            public Optional<U> getCompleted() {
+                return Deferrable.this.getCompleted()
+                    .map(mapper);
+            }
+
+            @Override
+            public U completeAndGet() {
+                return mapper.apply(Deferrable.this.completeAndGet());
+            }
+        };
+    }
+
+    /**
+     * Chains two {@link Deferrable}s.
+     *
+     * @param mapper A function which creates the next {@link Deferrable} from the result of the first one.
+     * Creating the invocation may be expensive, so this method avoids calling the mapper twice if possible.
+     */
+    default <U> Deferrable<U> flatMap(Function<? super T, Deferrable<U>> mapper) {
+        return getCompleted()
+            .map(mapper)
+            .orElseGet(() -> Deferrable.deferred(() -> mapper
+                .apply(Deferrable.this.completeAndGet())
+                .completeAndGet()));
+    }
+
+    /**
+     * An already completed result, can be successful or failed.
+     */
+    static <T> Deferrable<T> completed(T successfulResult) {
+        return new Deferrable<T>() {
+            @Override
+            public Optional<T> getCompleted() {
+                return Optional.of(successfulResult);
+            }
+
+            @Override
+            public T completeAndGet() {
+                return successfulResult;
+            }
+        };
+    }
+
+    /**
+     * An invocation with no pre-computed result, requiring to do the expensive computation on {@link #completeAndGet()}.
+     */
+    static <T> Deferrable<T> deferred(Supplier<T> result) {
+        return new Deferrable<T>() {
+            private transient volatile T value;
+
+            @Override
+            public Optional<T> getCompleted() {
+                return Optional.ofNullable(value);
+            }
+
+            @Override
+            public T completeAndGet() {
+                if (value == null) {
```

Comment:
It seems we don't specify the nullability contract regarding the `Deferrable`.

I would presume by default that `null` result is a valid result, and we do not want to recompute it.
As such, I would have a special private object to mark the state as unresolved and compare value to it instead of the null.

Regardless of the final implementation choice, we should also add the test for the case where the supplier returns null.

We can borrow more inspiration from Kotlin's Lazy implementation that does one more good thing: releasing the result supplier for GC, after it was used.
https://github.com/JetBrains/kotlin/blob/b388c7503b8f6c45473f79103bd270a19e82f13c/libraries/stdlib/jvm/src/kotlin/util/LazyJVM.kt#L55

Reply:
For nullability we actually define the behavior: all methods are assumed to be non-null by default because of this: https://github.com/gradle/gradle/commit/0833d2e96a14dd39dd026c9fe779f72b08a4b8b7

This is the behavior we are expecting in most of our functional stuff like `Try` as well. I wouldn't add specific tests for nullability either. In general we rely on `NPE`s and the IDE to spot nullability problems.

As for the releasing of the supplier, I'm not convinced it's worth doing it here. I understand why Kotlin would choose to do that for their very generic implementation, but they have very different needs than us. The Kotlin feature is part of the public API that needs to serve a ton of different usages, and cannot be modified easily. For us the functionality is not needed now, it can be easily added later, and reduced complexity is of value.

Reply:
Fair point, I forgot about the `NonNullApi` on the package level 👍